src/USER-ESFRICTION/fix_esfriction.cpp

```diff
@@ -0,0 +1,203 @@
+
+FixESFriction::FixESFriction(LAMMPS *lmp, int narg, char **arg) :
+  Fix(lmp, narg, arg)
+{
+  if (narg < 4) error->all(FLERR,"Illegal fix es-friction command");
+  // args -- 0: fix, 1: fix-id, 2: group-id
+  if (strcmp(update->unit_style,"metal") != 0) {
+    error->all(FLERR, "fix es-friction supports only metal units");
+  }
+  Z1 = utils::numeric(FLERR, arg[3], false, lmp); // atomic number of PKA
+  Z2 = utils::numeric(FLERR, arg[4], false, lmp); // atomic number of target
+  A1 = utils::numeric(FLERR, arg[5], false, lmp); // mass number of PKA
+  A2 = utils::numeric(FLERR, arg[6], false, lmp); // mass number of target
+  Ne_val = utils::numeric(FLERR, arg[7], false, lmp);  // no. of valance shell electrons
+  E_pka = utils::numeric(FLERR, arg[8], false, lmp);   // PKA energy
+  /* ES_cutoff is the energy cutoff for applying ES.
+     That is, atoms with energy below this will not experience ES */
+  ES_cutoff = utils::numeric(FLERR, arg[9], false, lmp);
+  // Compute total energy lost due to ES as per NRT model
+  double E_hat = get_ehat();
+  ES_loss_NRT = E_pka - E_hat;
+  // following computes coefficient of friction (beta) using LS model
+  double number_density = atom->natoms / 
+    (domain->xprd * domain->yprd * domain->zprd) * 1e30; // number-per-m3
+  double E_0 = h_cross*h_cross / (2*m_e) * 
+    pow( 3*pi*pi*Ne_val*(number_density), 2.0/3.0) * joules_to_ev; // eV
+  double v_0 = 5.931e5 * sqrt(E_0) * 1e-2; // ang/ps
+  // lambda is in eV-Angstrom-ps units
+  double lambda = 8.0 * pi * (e_charge * e_charge * 10) * bohr_radius * 
+    pow(Z1, 1.0/6.0) * Z1*Z2 / ( v_0 * pow( pow(Z1, 2.0/3.0) + 
+    pow(Z2, 2.0/3.0) , 3.0/2.0) ); 
+  beta = number_density * lambda / 1.0e30; //eV-ps-angstrom^-2 units
+  // Init energy lost due to ES as per LS model (calculated numerically)
+  ES_loss_LS = 0.0;
+  MPI_Comm_rank(world, &my_rank);
+  if (my_rank==0){
+    if(screen){
+      fprintf(screen, "E_pka = %g eV\n", E_pka);
+      fprintf(screen, "Energy loss due to ES (from NRT model) = %g\n", 
+        ES_loss_NRT);
+    }
+    if (logfile){
+      fprintf(logfile, "E_pka = %g eV\n", E_pka);
+      fprintf(logfile, "Energy loss due to ES (from NRT model) = %g\n", 
+        ES_loss_NRT);
+    }
+  }
+}
+
+/* ---------------------------------------------------------------------- */
+
+double FixESFriction::get_kenergy(double A, double vx, double vy, double vz)
+{
+  double gms_to_kg = 1e-3;  // grams to kg
+  double apps_to_mps = 100; // amstrong per picoseconds -to- m/s
+  return (0.5 * A * ( vx*vx + vy*vy + vz*vz ) * gms_to_kg * apps_to_mps * 
+    apps_to_mps * joules_to_ev / Nav); // in eV
+}
+
+/* ----------------------------------------------------------------------
+   Ehat is total energy lost due to ES based on NRT model 
+----------------------------------------------------------------------- */
+
+double FixESFriction::get_ehat()
+{
+  double Emax = 25.0 * pow(Z1, 4.0/3.0) * A1;
+  if (E_pka >= Emax){
```

Comment:
Please use `utils::logmesg` and `fmt::format` to simplify the code and protect this output with an if for `comm->me == 0` so that the warning is only printed once when running in parallel. Keep in mind that LAMMPS can be (and is) run across 1000s of MPI ranks.

Reply:
You may also consider using `Error::warning()`